segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3StorageTest.java

```diff
@@ -0,0 +1,265 @@
+                            Math.toIntExact(range.getFirst()));
+                    if ( bytesRead != range.getFirst() ) {
+                        throw new IllegalStateException("Unable to read from the object " + key);
+                    }
+                }
+                int bytesRead = ( (InputStream) content).read(totalByes, Math.toIntExact(range.getFirst()),
+                        Math.toIntExact(range.getLast() + 1 - range.getFirst()));
+
+                if ( bytesRead != range.getLast() + 1 - range.getFirst()) {
+                    throw new IllegalStateException("Not able to read from input stream.");
+                }
+                super.putObject( new PutObjectRequest(bucketName, key, (Object) new ByteArrayInputStream(totalByes)));
+                aclMap.put( key, aclMap.get(key).withSize(range.getLast() -1));
+            } catch (IOException e) {
+                throw new S3Exception("NoObject", 404, "NoSuchKey", key);
+            }
+        }
+
+        @Synchronized
+        @Override
+        public void setObjectAcl(String bucketName, String key, AccessControlList acl) {
+            AclSize retVal = aclMap.get(key);
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 404, "NoSuchKey", key);
+            }
+            aclMap.put(key, retVal.withAcl(acl));
+        }
+
+        @Synchronized
+        @Override
+        public void setObjectAcl(SetObjectAclRequest request) {
+            AclSize retVal = aclMap.get(request.getKey());
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 404, "NoSuchKey", request.getKey());
+            }
+            aclMap.put(request.getKey(), retVal.withAcl(request.getAcl()));
+        }
+
+        @Override
+        public AccessControlList getObjectAcl(String bucketName, String key) {
+            AclSize retVal = aclMap.get(key);
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 404, "NoSuchKey", key);
+            }
+            return retVal.getAcl();
+        }
+
+        public CopyPartResult copyPart(CopyPartRequest request) {
+            if ( aclMap.get(request.getKey()) == null ) {
+                throw new S3Exception("NoObject", 404, "NoSuchKey", request.getKey());
+            }
+
+            Range range = request.getSourceRange();
+            if ( range.getLast() == -1 ) {
+                range = Range.fromOffsetLength(0, aclMap.get(request.getSourceKey()).getSize());
+                request.withSourceRange(range);
+            }
+            CopyObjectResult result = executeRequest(client, request, CopyObjectResult.class);
+            CopyPartResult retVal = new CopyPartResult();
+            retVal.setPartNumber(request.getPartNumber());
+            retVal.setETag(result.getETag());
+            return retVal;
+        }
+
+        @Override
+        public void deleteObject(String bucketName, String key) {
```

Comment:
Should this and deleteObjects be synchronized too? If there is a concurrent delete between the get and the put in putObject it would be weird, but I'm not clear enough on the expected semantics to know if it would be wrong.